jina/executors/indexers/keyvalue.py

```diff
@@ -56,6 +63,53 @@ def close(self):
             """Close the file."""
             self._body.close()
 
+    def __getstate__(self):
+        # called on pickle save
+        if self.delete_on_dump:
+            self._delete_invalid_indices()
+        d = super().__getstate__()
+        return d
+
+    def _delete_invalid_indices(self):
+        if self.query_handler:
+            self.query_handler.close()
+        if self.write_handler:
+            self.write_handler.flush()
+            self.write_handler.close()
+
+        keys = []
+        vals = []
+        # we read the valid values and write them to the intermediary file
+        read_handler = self.ReadHandler(self.index_abspath, self.key_length)
```

Comment:
use read handler as context manager for better security

Reply:
It's not a context manager. It's a custom class, without the `__enter__` / `__exit__`

Reply:
so maybe it would be good to have it work as such for better exception handling

Reply:
Good idea. Working on it

Reply:
Added it